### PR TITLE
retry sync use GTID if use GTID sync at the beginning

### PR DESCRIPTION
**Pull request by WangXiangUSTC**
retry sync use GTID if use GTID sync at the beginning
1. add updateGTIDSet function of GTIDSet interface
2. update GTIDSet when meet GTID EVENT

### CODE REVIEW

File: replication/binlogsyncer.go [hunk 1]
```diff
@@ -77,6 +78,10 @@ type BinlogSyncer struct {
 
 	nextPos Position
 
+	useGTID bool
```
Comment (siddontang) [inline replication/binlogsyncer.go]: seem you don't need useGTID here, you can use gset == nil to check whether syncs with gtid or not directly.

Comment (GregoryIan) [inline replication/binlogsyncer.go]: it's not all right.  if u want sync by GTID, u must have initial GTID set and incremental GTID set information.

File: replication/binlogsyncer.go [hunk 2]
```diff
@@ -517,8 +554,8 @@ func (b *BinlogSyncer) onStream(s *BinlogStreamer) {
 			log.Error(err)
 
 			// we meet connection error, should re-connect again with
-			// last nextPos we got.
-			if len(b.nextPos.Name) == 0 {
+			// last nextPos or nextGTID we got.
+			if len(b.nextPos.Name) == 0 && b.gset == nil {
```
Comment (GregoryIan) [inline replication/binlogsyncer.go]: seem it's not right

**Comment by GregoryIan** [comment]
can we put `gset` in events(like XIDEvent)?  @siddontang   if so, `syncer` don't need handle gtid event, same to other projects

**Comment by siddontang** [comment]
You can file a PR to support it. @GregoryIan

File: replication/binlogsyncer.go [hunk 3]
```diff
@@ -593,8 +630,20 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
 		b.nextPos.Name = string(re.NextLogName)
 		b.nextPos.Pos = uint32(re.Position)
 		log.Infof("rotate to %s", b.nextPos)
+	} else if  ge, ok := e.Event.(*GTIDEvent); b.useGTID && ok {
```
Comment (holys) [inline replication/binlogsyncer.go]: How about to use `switch` here ?

**Comment by holys** [comment]
The rest of them LGTM

File: replication/binlogsyncer.go (hunk 3 above)
Comment (siddontang) [inline replication/binlogsyncer.go]: seem you can update gtid all time no matter enabling useGTID or not.

**Comment by GregoryIan** [comment]
LGTM

File: replication/binlogsyncer.go (hunk 2 above)
Comment (GregoryIan) [inline replication/binlogsyncer.go]: maybe `if len(b.nextPos.Name ==0)` if enough

File: replication/binlogsyncer.go [hunk 4]
```diff
@@ -588,11 +625,29 @@ func (b *BinlogSyncer) parseEvent(s *BinlogStreamer, data []byte) error {
...
-	if re, ok := e.Event.(*RotateEvent); ok {
-		b.nextPos.Name = string(re.NextLogName)
-		b.nextPos.Pos = uint32(re.Position)
+	switch event := e.Event.(type) {
+	case *RotateEvent:
+		b.nextPos.Name = string(event.NextLogName)
+		b.nextPos.Pos = uint32(event.Position)
 		log.Infof("rotate to %s", b.nextPos)
+	case *GTIDEvent:
+		if !b.useGTID {
```
Comment (siddontang) [inline replication/binlogsyncer.go]: why not update GITD all the time?

**Comment by WangXiangUSTC** [comment]
@siddontang do you think we should keep the useGTID? and update GTID set all the time?

**Comment by siddontang** [comment]
@WangXiangUSTC yes

So we can add the GTID in the Event like position all the time. /cc @GregoryIan

**Comment by WangXiangUSTC** [comment]
but if user dosen't use gtid sync data, I can't initial the gtid set. Maybe we can just update gtid set when useGTID==True now, and update it all the time when we really need later.

File: mysql/gtid.go [hunk 5]
```diff
@@ -11,6 +11,8 @@ type GTIDSet interface {
 	Equal(o GTIDSet) bool
 
 	Contain(o GTIDSet) bool
+
+	UpdateGTIDSet(GTIDStr string) error
```
Comment (siddontang) [inline mysql/gtid.go]: s/UpdateGTIDSet/Update/

**Comment by WangXiangUSTC** [comment]
done @siddontang

**Pull request merged**